### PR TITLE
Separate local and remote concepts

### PR DESCRIPTION
**Pull request by adammitchelldev**
## What is the goal of this PR?

To separate local and remote concepts, so that a user must provide a transaction to a local concept in order to access remote methods. This is the new pattern implemented by other clients.

This also allows us to make use of pre-filled results from queries.

## What are the changes implemented in this PR?

* Refactor concepts into Javascript classes (rather than the old-style Object construction.)
* Add a local concept test to cover the local concept factory and the concept methods

### CODE REVIEW

File: src/service/Session/concept/ConceptFactory.js [hunk 1]
```diff
@@ -17,115 +17,60 @@
...
+    default:
+      throw "BaseType not recognised.";
+  }
+}
 
-module.exports = ConceptFactory;
+ConceptFactory.createLocalConcept = function(grpcConcept) {
+  const baseType = BaseTypeConstants.fromGrpcConcept(grpcConcept);
+  switch (baseType) {
+    case BaseType.ENTITY: return new LocalConcept.Thing(grpcConcept);
```
Comment (flyingsilverfin) [inline src/service/Session/concept/ConceptFactory.js:64]: Is an `Entity` supposed to return a local `Thing`?

File: src/service/Session/concept/ConceptFactory.js [hunk 2]
```diff
@@ -17,115 +17,60 @@
...
-  Methods.rule
-);
-
-const metaschemaProto = Object.assign({},
-  Methods.concept,
-  Methods.schemaConcept,
-);
+ConceptFactory.createRemoteConcept = function(conceptId, baseType, txService) {
+  switch (baseType) {
+    case BaseType.ENTITY: return new RemoteConcept.Thing(conceptId, BaseType.ENTITY, txService);
```
Comment (flyingsilverfin) [inline src/service/Session/concept/ConceptFactory.js:47]: `Entity` -> `Thing`?

File: src/service/Session/concept/LocalConcept.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+const BaseType = Constant.baseType;
+const ConceptFactory = require('./ConceptFactory');
+const ProtoDataType = require("../../../../grpc/nodejs/protocol/session/Concept_pb").AttributeType.DATA_TYPE;
+
+function convertDataType(dataTypeRes) {
+    if (dataTypeRes.hasNull()) return null;
+    switch (dataTypeRes.getDatatype()) {
+        case ProtoDataType.STRING: return "String";
+        case ProtoDataType.BOOLEAN: return "Boolean";
+        case ProtoDataType.INTEGER: return "Integer";
```
Comment (flyingsilverfin) [inline src/service/Session/concept/LocalConcept.js:32]: we don't need Long and Integer anymore I believe, same with Float and Double? What do other clients support atm?

File: tests/service/session/transaction/Query.test.js [hunk 4]
```diff
@@ -26,31 +26,41 @@ let tx;
...
+    const iterator = await tx.query('match $y isa person, has identifier $x; get $x;');
+    const response = await iterator.next();
+    const id = response.get('x');
+    expect(id.value()).toEqual('a');
+    expect(id.dataType()).toEqual('String');
+    expect(id.type().label()).toEqual('identifier');
+    expect(id.isInferred()).toEqual(false);
+    expect(id.type().isImplicit()).toEqual(false);
+  })
+
```
Comment (flyingsilverfin) [inline tests/service/session/transaction/Query.test.js:65]: I believe you're missing the new test for duplicate role players that is in the open PR that is meant to be included in this one :)

Comment (flyingsilverfin) [reply tests/service/session/transaction/Query.test.js:65]: ahh was that included in a separate PR?

File: src/service/Session/concept/ConceptFactory.js (hunk 1 above)
Comment (adammitchelldev) [reply src/service/Session/concept/ConceptFactory.js:64]: Yes, there is no `Entity` class, this was how the code was before too. We can change it of course.

File: src/service/Session/concept/LocalConcept.js (hunk 3 above)
Comment (adammitchelldev) [reply src/service/Session/concept/LocalConcept.js:32]: They are still in the protocol so we should remove them at the same time we remove them from the protocol.

File: src/service/Session/concept/ConceptFactory.js (hunk 1 above)
Comment (flyingsilverfin) [reply src/service/Session/concept/ConceptFactory.js:64]: Would look nicer i think!

File: tests/service/session/transaction/Query.test.js (hunk 4 above)
Comment (adammitchelldev) [reply tests/service/session/transaction/Query.test.js:65]: Is it meant to be included in `Query.test.js`?

Comment (flyingsilverfin) [reply tests/service/session/transaction/Query.test.js:65]: no, should be in `Relation` tests

**Comment by flyingsilverfin** [comment]
There are some other bits of the tests that haven't been updated with the changes from https://github.com/graknlabs/client-nodejs/pull/90/files:

renaming `set` -> `array`, once removing an `Array.fromSet()` etc.

File: tests/service/session/transaction/Query.test.js (hunk 4 above)
Comment (adammitchelldev) [reply tests/service/session/transaction/Query.test.js:65]: Got them in now.

File: src/service/Session/util/GrpcIteratorFactory.js [hunk 5]
```diff
@@ -72,7 +72,8 @@ class Iterator {
 
   async _nextBatch(iterRequest) {
     this._iterator = await this._communicator.iterateUntil(RequestBuilder.txIter(iterRequest), (res) => {
-      return res.getIterRes().getDone() == true;
+      const iterRes = res.getIterRes();
+      return (iterRes.getDone() == true) || (iterRes.getIteratorid() > 0);
```
Comment (flyingsilverfin) [inline src/service/Session/util/GrpcIteratorFactory.js:76]: is `Id` not capitalised :P
